File: backend/src/main/java/me/performancereservation/domain/performance/mapper/PerformanceMapper.java
package me.performancereservation.domain.performance.mapper;

import me.performancereservation.domain.performance.dto.performance.request.PerformanceCreateRequest;
import me.performancereservation.domain.performance.dto.performance.response.PerformanceDetailResponse;
import me.performancereservation.domain.performance.dto.performance.response.PerformancePageResponse;
import me.performancereservation.domain.performance.dto.performance.response.PerformanceManagerDetailResponse;
import me.performancereservation.domain.performance.dto.performance.response.PerformanceManagerPageResponse;
import me.performancereservation.domain.performance.dto.performanceschedule.PerformanceScheduleResponse;
import me.performancereservation.domain.performance.entities.Performance;
import me.performancereservation.domain.performance.entities.PerformanceSchedule;
import me.performancereservation.domain.performance.enums.PerformanceCategory;
import me.performancereservation.domain.performance.enums.PerformanceStatus;

import java.util.List;

public class PerformanceMapper {

    /** 요청 객체 엔티티 변환
     *
     * 파일 아이디, 공연자 아이디 추가 필요
     * @param request
     * @return Performance
     */
    public static Performance toEntity(PerformanceCreateRequest request, Long managerId) {
        return Performance.builder()
                .fileId(request.fileId())
                .managerId(managerId)
                .title(request.title())
                .venue(request.venue())
                .price(request.price())
                .totalSeats(request.totalSeats())
                .category(PerformanceCategory.valueOf(request.category()))
                .startDate(request.startDate())
                .endDate(request.endDate())
                .description(request.description())
                .status(PerformanceStatus.PENDING)
                .build();
    }

    /** 공연 상세 페이지용 응답
     *
     * @param performance
     * @param fileUrl
     * @param schedules
     * @return PerformanceDetailResponse
     */
    public static PerformanceDetailResponse toDetailResponse(Performance performance, String fileUrl, List<PerformanceSchedule> schedules) {
        return new PerformanceDetailResponse(
                performance.getId(),
                performance.getTitle(),
                performance.getPrice(),
                performance.getTotalSeats(),
                performance.getVenue(),
                performance.getDescription(),
                performance.getStatus().toString(),
                fileUrl,
                schedules.stream().map(PerformanceScheduleMapper::toResponse).toList()
        );
    }

    /** 공연 목록 페이지 응답용
     *
     * @param performance
     * @param fileUrl
     * @return PerformanceListResponse
     */
    public static PerformancePageResponse toListResponse(Performance performance, String fileUrl) {
        return new PerformancePageResponse(
                performance.getId(),
                fileUrl,
                performance.getTitle(),
                performance.getPrice(),
                performance.getStartDate(),
                performance.getVenue()
        );
    }

    /** 공연자 공연 목록 페이지 응답용
     *
     * @param performance
     * @param fileUrl
     * @return PerformanceManagerListResponse
     */
    public static PerformanceManagerPageResponse toManagerListResponse(Performance performance, String fileUrl) {
        return new PerformanceManagerPageResponse(
                performance.getId(),
                fileUrl,
                performance.getTitle(),
                performance.getStartDate(),
                performance.getVenue(),
                performance.getStatus().toString()
        );
    }

    /** 공연자 공연 관리 페이지 응답용
     *
     * @param performance
     * @param fileUrl
     * @return PerformanceManagerDetailResponse
     */
    public static PerformanceManagerDetailResponse toManagerDetailResponse(Performance performance,
                                                                           String fileUrl,
                                                                           List<PerformanceScheduleResponse> schedules) {
        return new PerformanceManagerDetailResponse(
                performance.getId(),
                fileUrl,
                performance.getTitle(),
                performance.getVenue(),
                performance.getStatus().toString(),
                performance.getTotalSeats(),
                schedules
        );
    }
}

File: backend/src/main/kotlin/me/performancereservation/domain/performance/mapper/PerformanceMapper.kt
package me.performancereservation.domain.performance.mapper

import me.performancereservation.domain.performance.dto.performance.*
import me.performancereservation.domain.performance.dto.performanceschedule.PerformanceScheduleResponse
import me.performancereservation.domain.performance.entities.Performance
import me.performancereservation.domain.performance.entities.PerformanceSchedule
import me.performancereservation.domain.performance.enums.PerformanceCategory
import me.performancereservation.domain.performance.enums.PerformanceStatus

object PerformanceMapper {
    /** 요청 객체 엔티티 변환
     *
     * @param request
     * @return Performance
     */
    @JvmStatic
    fun toEntity(request: PerformanceCreateRequest, managerId: Long): Performance {
        return Performance(
            fileId = request.fileId,
            managerId = managerId,
            title = request.title,
            venue = request.venue,
            price = request.price,
            totalSeats = request.totalSeats,
            category = PerformanceCategory.valueOf(request.category),
            startDate = request.startDate,
            endDate = request.endDate,
            description = request.description,
            status = PerformanceStatus.PENDING)

    }

    /** 공연 상세 페이지용 응답
     *
     * @param performance
     * @param fileUrl
     * @param schedules
     * @return PerformanceDetailResponse
     */
    @JvmStatic
    fun toDetailResponse(
        performance: Performance,
        fileUrl: String,
        bookmarked: Boolean,
        schedules: List<PerformanceSchedule>
    ): PerformanceDetailResponse {
        return PerformanceDetailResponse(
            id = performance.id!!,
            title = performance.title,
            price = performance.price,
            totalSeats = performance.totalSeats,
            venue = performance.venue,
            description = performance.description,
            status = performance.status,
            fileUrl = fileUrl,
            startDate = performance.startDate,
            endDate = performance.endDate,
            bookmarked = bookmarked,
            schedules = schedules.stream().map { schedule: PerformanceSchedule ->
                PerformanceScheduleMapper.toResponse(schedule)
            }.toList()
        )
    }

    /** 공연 목록 페이지 응답용
     *
     * @param performance
     * @param fileUrl
     * @return PerformanceListResponse
     */
    @JvmStatic
    fun toListResponse(performance: Performance, fileUrl: String?): PerformancePageResponse {
        return PerformancePageResponse(
            id = performance.id!!,
            fileUrl = fileUrl,
            title = performance.title,
            price = performance.price,
            startDate = performance.startDate,
            endDate = performance.endDate,
            venue = performance.venue,
            category = performance.category
        )
    }

    /** 공연자 공연 목록 페이지 응답용
     *
     * @param performance
     * @param fileUrl
     * @return PerformanceManagerListResponse
     */
    @JvmStatic
    fun toManagerListResponse(performance: Performance, fileUrl: String): PerformanceManagerPageResponse {
        return PerformanceManagerPageResponse(
            id = performance.id!!,
            fileUrl = fileUrl,
            title = performance.title,
            startDate = performance.startDate,
            endDate = performance.endDate,
            venue = performance.venue,
            status = performance.status,
            category = performance.category
        )
    }

    /** 공연자 공연 관리 페이지 응답용
     *
     * @param performance
     * @param fileUrl
     * @return PerformanceManagerDetailResponse
     */
    @JvmStatic
    fun toManagerDetailResponse(
        performance: Performance,
        fileUrl: String,
        schedules: List<PerformanceScheduleResponse>
    ): PerformanceManagerDetailResponse {
        return PerformanceManagerDetailResponse(
            id =performance.id!!,
            fileUrl = fileUrl,
            title = performance.title,
            venue = performance.venue,
            status = performance.status,
            totalSeats = performance.totalSeats,
            startDate = performance.startDate,
            endDate = performance.endDate,
            description = performance.description,
            category = performance.category,
            schedules = schedules
        )
    }
}
